File: src/ast.ml

type t =
  | Null
  | Int of t 
  | Const of int
  | Ng of t
  | Var of string
  | Op of operator
  | Func of func

and operator =
  | Plus of t * t
  | Minus of t * t
  | Time of t * t
  | Divid of t * t

and func =
  | Pow of t * t
  | Sqrt of t
  | Exp of t
  | Ln of t
  | Sin of t
  | Asin of t
  | Cos of t
  | Acos of t
  | Tan of t
  | Atan of t
[@@deriving show]

File: Ast.fs
module Ast

type T =
    | Null
    | Int of T 
    | Const of int
    | Ng of T
    | Var of string
    | Op of Operator
    | Func of Func

and Operator =
    | Plus of T * T
    | Minus of T * T
    | Time of T * T
    | Divid of T * T

and Func =
    | Pow of T * T
    | Sqrt of T
    | Exp of T
    | Ln of T
    | Sin of T
    | Asin of T
    | Cos of T
    | Acos of T
    | Tan of T
    | Atan of T